File: sources/CPP/include/Ers/UI/Widgets/LicensingWidget.cpp
#include "LicensingWidget.h"

#include "Ers/Api.h"

namespace Ers
{
    LicensingWidget::LicensingWidget()
    {
        corePtr = Ers::Engine::ERS_LicensingWidget_Create();
    }

    LicensingWidget::~LicensingWidget()
    {
        Ers::Engine::ERS_LicensingWidget_Destroy(corePtr);
    }

    void LicensingWidget::Window(const char* name, bool* open, ImGuiWindowFlags flags)
    {
        Ers::Engine::ERS_LicensingWidget_Window(corePtr, name, open, flags);
    }

    void LicensingWidget::Widget()
    {
        Ers::Engine::ERS_LicensingWidget_Widget(corePtr);
    }
} // namespace Ers

File: sources/CSharp/src/Ers/UI/Widgets/LicensingWidget.cs
using Ers.Engine;
using ImGuiNET;

namespace Ers
{
    /// <summary>
    /// Widget for license activation and management.
    /// </summary>
    public class LicensingWidget : IDisposable
    {
        /// <summary>
        /// Native pointer to the core instance.
        /// </summary>
        public IntPtr CorePtr;

        /// <summary>
        /// Constructs a new <see cref="LicensingWidget"/>.
        /// </summary>
        public LicensingWidget() { CorePtr = ErsEngine.ERS_LicensingWidget_Create(); }

        /// <summary>
        /// Destroy the widget instance.
        /// </summary>
        public void Dispose()
        {
            if (CorePtr != IntPtr.Zero)
            {
                ErsEngine.ERS_LicensingWidget_Destroy(CorePtr);
                CorePtr = IntPtr.Zero;
            }
        }

        /// <summary>
        /// Show the window.
        /// </summary>
        /// <param name="name">The name for the window.</param>
        public void Window(string name)
        {
            unsafe
            {
                fixed(byte* utf8Name = name.ToUtf8NullTerminated())
                {
                    ErsEngine.ERS_LicensingWidget_Window(CorePtr, utf8Name, null, (int)ImGuiWindowFlags.None);
                }
            }
        }

        /// <summary>
        /// Show the window.
        /// </summary>
        /// <param name="name">The name for the window.</param>
        /// <param name="open">Whether the window is open.</param>
        /// <param name="flags">Any <see cref="ImGuiWindowFlags"/> for the window.</param>
        public void Window(string name, ref bool open, ImGuiWindowFlags flags = ImGuiWindowFlags.None)
        {
            unsafe
            {
                fixed(byte* utf8Name = name.ToUtf8NullTerminated()) fixed(bool* openPtr = &open)
                {
                    ErsEngine.ERS_LicensingWidget_Window(CorePtr, utf8Name, openPtr, (int)flags);
                }
            }
        }

        /// <summary>
        /// Show the widget.
        /// </summary>
        public void Widget() => ErsEngine.ERS_LicensingWidget_Widget(CorePtr);
    }
}
